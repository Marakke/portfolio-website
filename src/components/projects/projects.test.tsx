import React from "react";
import { render, fireEvent } from "@testing-library/react";
import * as Gatsby from "gatsby";
import Projects from "./index";

const useStaticQuery = jest.spyOn(Gatsby, `useStaticQuery`);

const mockUseStaticQuery = {
  projects: {
    title: "Test Title",
    descriptionTop: "Test Description Top",
    descriptionBottom: "Test Description Bottom",
    buttonText: "Button Text",
    buttonLink: "Button Link",
    projects: [
      {
        client: "Client",
        name: "Project Name",
        link: "Project Link",
        tagHeader: [
          {
            name: "Tag Header"
          }
        ],
        tagFooter: [
          {
            name: "Tag Footer"
          }
        ],
      }
    ]
  }
};

describe("Projects", () => {
  beforeEach(() => {
    useStaticQuery.mockImplementation(() => mockUseStaticQuery);
    window.open = jest.fn();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders the Projects component correctly", () => {
    const { container, getByTestId, getByRole } = render(<Projects />);
    expect(container).toMatchSnapshot();

    const projectsContainer = getByTestId("projects-container");
    expect(projectsContainer).toBeInTheDocument();

    const titleElement = getByRole("heading", { name: "Test Title" });
    expect(titleElement).toBeInTheDocument();

    const paragraphElements = container.querySelectorAll("p");
    expect(paragraphElements.length).toEqual(11);

    expect(paragraphElements[0].textContent).toBe("Test Description Top");
    expect(paragraphElements[1].textContent).toBe("Test Description Bottom");

    expect(paragraphElements[2].textContent).toBe("Client");
    expect(paragraphElements[3].textContent).toBe("Project Name");
    expect(paragraphElements[4].textContent).toBe("Tag Header");
    expect(paragraphElements[5].textContent).toBe("Tag Footer");
  });

  it("clicks project button and opens new tab", () => {
    const { container } = render(<Projects />);

    const buttonElement = container.querySelectorAll("button");
    expect(buttonElement[1]).toBeInTheDocument();

    fireEvent.click(buttonElement[1]);

    expect(window.open).toHaveBeenCalledWith("Project Link", "_blank");
  });

  it("clicks GitHub button and opens new tab", () => {
    const { container } = render(<Projects />);

    const buttonElement = container.querySelectorAll("button");
    expect(buttonElement[0]).toBeInTheDocument();

    fireEvent.click(buttonElement[0]);

    expect(window.open).toHaveBeenCalledWith("Button Link", "_blank");
  });
});
