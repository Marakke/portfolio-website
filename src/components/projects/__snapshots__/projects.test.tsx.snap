// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects renders the Projects component correctly 1`] = `
<div>
  <div
    class="projects-container"
    data-testid="projects-container"
    id="projects"
  >
    <h2>
      Test Title
    </h2>
    <div
      class="projects-content"
    >
      <div
        class="content-left"
      >
        <div>
          <p
            class="description"
          >
            Test Description Top
          </p>
          <p
            class="description"
          >
            Test Description Bottom
          </p>
        </div>
        <button
          class="buttonDesktop"
        >
          <img
            alt="GitHub Icon"
            src="test-file-stub"
          />
          Button Text
        </button>
      </div>
      <div
        class="content-right"
      >
        <button
          class="project "
        >
          <p
            class="client"
          >
            Client
          </p>
          <p
            class="name"
          >
            Project Name
          </p>
          <div
            class="tags"
          >
            <div
              class="tagsHeader"
            >
              <p
                class="tagHeader"
              >
                Tag Header
              </p>
            </div>
            <div
              class="tagsFooter"
            >
              <p
                class="tagFooter"
              >
                Tag Footer
              </p>
            </div>
          </div>
        </button>
      </div>
      <button
        class="buttonMobile"
      >
        <img
          alt="GitHub Icon"
          src="test-file-stub"
        />
        Button Text
      </button>
    </div>
    <div
      class="lines horizontal projects"
    >
      <p
        class="line horizontal"
      />
      <p
        class="line horizontal"
      />
      <p
        class="line horizontal"
      />
      <p
        class="line horizontal"
      />
      <p
        class="line horizontal"
      />
    </div>
  </div>
</div>
`;
